service/service_test.go

```diff
@@ -123,21 +122,24 @@ func TestService_Recovery(t *testing.T) {
 	s.openRoundMutex.Lock()
 	req.Nil(s.openRound)
 	s.openRoundMutex.Unlock()
+	s.Lock()
 	req.Equal(len(s.executingRounds), 0)
-
+	s.Unlock()
 	// Create a new service instance.
-	sig = signal.NewSignal()
-	s, err = NewService(sig, cfg, tempdir)
+	s, err = NewService(cfg, tempdir)
 	req.NoError(err)
 
 	err = s.Start(broadcaster)
 	req.NoError(err)
 
 	// Service instance should recover 2 rounds: round 1 in executing state, and round 2 in open state.
-	req.Equal(1, len(s.executingRounds))
+	req.Eventually(func() bool { s.Lock(); defer s.Unlock(); return len(s.executingRounds) == 1 }, time.Second, time.Millisecond)
+	s.Lock()
 	first, ok := s.executingRounds["0"]
+	s.Unlock()
 	req.True(ok)
-	req.Equal(s.openRoundID(), "1")
+	req.NotNil(s.openRoundID())
+	req.Equal(*s.openRoundID(), "1")
```

Comment:
This calls `s.openRound()` twice. What's the reason for returning a pointer instead o the value?

If it we stick with the pointer this should avoid a panic and calling `openRoundID()` twice.

```suggestion
	openRoundID := s.openRoundID()
	req.NotNil(openRoundID)
	req.Equal("1", *openRoundID)
```

Also the first parameter is `expected` and the second `actual` 😉. Otherwise this might lead to confusing messages when the assert fails.

Reply:
Calling `openRoundID()` twice is not that big of a deal, it's just a getter.

I made it return a pointer because `s.openRound` **is not always set**. It's set after `Service::Start()` returns. It's a problem in the async API. An RPC call `GetInfo()` could panic because of it (a segfault).

Anyway - I rewritten `Start()` to set `Service::started bool` only after the Service is completely started (and `Service::openround` is set) and reverted to returning a `string` here, which is safe now.